api/src/main/java/org/openmrs/api/db/hibernate/search/LuceneAnalyzerFactory.java

```diff
@@ -0,0 +1,47 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at http://mozilla.org/MPL/2.0/. OpenMRS is also distributed under
+ * the terms of the Healthcare Disclaimer located at http://openmrs.org/license.
+ *
+ * Copyright (C) OpenMRS Inc. OpenMRS is a registered trademark and the OpenMRS
+ * graphic logo is a trademark of OpenMRS Inc.
+ */
+package org.openmrs.api.db.hibernate.search;
+
+import org.apache.lucene.analysis.core.KeywordTokenizerFactory;
+import org.apache.lucene.analysis.core.LowerCaseFilterFactory;
+import org.apache.lucene.analysis.core.WhitespaceTokenizerFactory;
+import org.apache.lucene.analysis.ngram.EdgeNGramFilterFactory;
+import org.apache.lucene.analysis.ngram.NGramFilterFactory;
+import org.apache.lucene.analysis.standard.ClassicFilterFactory;
+import org.hibernate.search.annotations.Factory;
+import org.hibernate.search.cfg.SearchMapping;
+
+public class LuceneAnalyzerFactory {
+	@Factory
+	public SearchMapping getSearchMapping() {
+		SearchMapping mapping = new SearchMapping();
+		mapping
+			.analyzerDef(LuceneAnalyzers.PHRASE_ANALYZER, KeywordTokenizerFactory.class)
+			.filter(ClassicFilterFactory.class)
+			.filter(LowerCaseFilterFactory.class);
+		mapping.analyzerDef(LuceneAnalyzers.EXACT_ANALYZER, WhitespaceTokenizerFactory.class)
+			.filter(ClassicFilterFactory.class)
+			.filter(LowerCaseFilterFactory.class);
+		mapping.analyzerDef(LuceneAnalyzers.START_ANALYZER, WhitespaceTokenizerFactory.class)
+			.filter(ClassicFilterFactory.class)
+			.filter(LowerCaseFilterFactory.class)
+			.filter(EdgeNGramFilterFactory.class)
+                .param("minGramSize", "2")
+                .param("maxGramSize", "20");
+		mapping.analyzerDef(LuceneAnalyzers.ANYWHERE_ANALYZER, WhitespaceTokenizerFactory.class)
+			.filter(ClassicFilterFactory.class)
+			.filter(LowerCaseFilterFactory.class)
+			.filter(NGramFilterFactory.class)
+			.param("minGramSize", "2")
+			.param("maxGramSize", "20");
+		return mapping;
```

Comment:
This looks right... I wonder if can come up with a good strategy for testing that things are still working correctly?

Reply:
Forgive my ignorance of Lucene, but is the idea that we define the analyzers here and then the annotations on the actual domain objects (like PersonName) reference these analyzers?  Can you mix these two ways of configuring?  (ie setting up the analyzers programmatically but then referencing them via annotations?)

Reply:
Good question! Yes, PersonName is referencing the analyzers using field annotations: see [PersonName:58](https://github.com/openmrs/openmrs-core/blob/640cd7dd47f65c02574458f4b25d8b114aa817ed/api/src/main/java/org/openmrs/PersonName.java#L58). I have read that you can't define analyzers both with annotations and programmatically, though.

Reply:
So we are saying here "according to the docs, this shouldn't work, but it looks like it actually is?".  I'd be a little concerned with that which could hopefully be mitigated by adding some tests as @mseaton mentions.

Reply:
No, I think the distinction between defining and using analyzers is important here.

> Today, the programmatic API cannot be used on a class annotated with Hibernate Search annotations, chose one approach or the other.

We're only using the programmatic API to define analyzers, we're not "using it on a class." Classes are continuing to use analyzers via annotations (`@Field`).

Reply:
Great, thanks for the clarification @brandones

Reply:
Good to know. I thought that all needs to be in the programmatic API. In the long run we still may want to move from annotations on a class to programmatic API, but it's okay to do it for analyzers only, if it works.